internal/services/cdn/cdn_frontdoor_helpers.go

```diff
@@ -339,76 +383,265 @@ func updateRouteAssociations(d *pluginsdk.ResourceData, meta interface{}, routeI
+		}
 	}
 
-	// Make sure all of the custom domains that are referenced by the resource are actually associated with the route...
-	notAssociated := make([]string, 0)
+	return nil, fmt.Errorf("unable to parse CDN FrontDoor Custom Domain: no value was passed")
+}
+
+// Checks to make sure the list of CDN FrontDoor Custom Domains does not contain duplicate entries
+func sliceHasDuplicates(input []interface{}, resourceTxt string) error {
+	k := make(map[string]bool)
+	if len(input) == 0 || input == nil {
+		return nil
+	}
+
+	for _, v := range input {
+		if _, d := k[strings.ToLower(v.(string))]; !d {
+			k[strings.ToLower(v.(string))] = true
+		} else {
+			return fmt.Errorf("duplicate %[1]s detected, please remove all duplicate entries for the %[1]s(ID: %q) from your configuration block", resourceTxt, v.(string))
+		}
+	}
+
+	return nil
+}
 
-	for _, v := range resourceCustomDomains {
-		customDomain, err := parse.FrontDoorCustomDomainIDInsensitively(v.(string))
+func routeSliceHasDuplicates(input *[]parse.FrontDoorRouteId, resourceName string) error {
+	k := make(map[string]bool)
+	if len(*input) == 0 || input == nil {
+		return nil
+	}
+
+	for _, route := range *input {
+		if _, d := k[strings.ToLower(route.ID())]; !d {
+			k[strings.ToLower(route.ID())] = true
+		} else {
+			return fmt.Errorf("duplicate %[1]s detected, please remove all duplicate entries for the %[1]s(ID: %q) from your configuration block", resourceName, route.ID())
+		}
+	}
+
+	return nil
+}
+
+// Determines what CDN FrontDoor Routes need to be removed/disassociated from this CDN FrontDoor Custom Domain
+func routeDelta(oldRoutes *[]parse.FrontDoorRouteId, newRoutes *[]parse.FrontDoorRouteId) (*[]parse.FrontDoorRouteId, *[]parse.FrontDoorRouteId) {
+	remove := make([]parse.FrontDoorRouteId, 0)
+	shared := make([]parse.FrontDoorRouteId, 0)
+	if len(*newRoutes) == 0 || newRoutes == nil {
+		return oldRoutes, &shared
+	}
+
+	if len(*oldRoutes) == 0 || oldRoutes == nil {
+		return &remove, &shared
+	}
+
+	// just find what old routes are not in the new route list...
+	for _, oldRoute := range *oldRoutes {
+		if !routeSliceContains(newRoutes, oldRoute.ID()) {
+			remove = append(remove, oldRoute)
+		} else {
+			shared = append(shared, oldRoute)
+		}
+	}
+
+	return &remove, &shared
+}
+
+func normalizeRuleSetIds(input []interface{}) ([]interface{}, error) {
+	out := make([]interface{}, 0)
+	if len(input) == 0 || input == nil {
+		return out, nil
+	}
+
+	for _, ruleSet := range input {
+		id, err := parse.FrontDoorRuleSetIDInsensitively(ruleSet.(string))
 		if err != nil {
-			return fmt.Errorf("unable to parse %q: %+v", v.(string), err)
+			return nil, err
 		}
 
-		if !sliceContainsString(routeCustomDomains, customDomain.ID()) {
-			notAssociated = append(notAssociated, fmt.Sprintf("%q", customDomain.ID()))
+		out = append(out, id.ID())
+	}
+
+	return out, nil
+}
+
+func routesInsensitively(input []interface{}) (*[]parse.FrontDoorRouteId, []interface{}, error) {
```

Comment:
rather than using this for Expand and Flatten, can we split this function? the Expand should always be case-sensitive whereas the Flatten can be insensitive (to account for the API)

Reply:
Fixed.